backend/tests/test_streams.py

```diff
@@ -0,0 +1,183 @@
+"""Test streams."""
+
+import asyncio
+
+import pytest
+from gatterserver.streams import StreamId, StreamManager, StreamPacket
+
+
+def test_stream_id_type():
+    s = StreamId(device_id=4, stream_id=8)
+    assert s
+    assert s.device_id == 4
+    assert s.stream_id == 8
+
+    with pytest.raises(ValueError):
+        s = StreamId(device_id=4)
+
+    with pytest.raises(ValueError):
+        s = StreamId(stream_id=4)
+
+    with pytest.raises(ValueError):
+        s = StreamId(device_id=-1)
+
+    with pytest.raises(ValueError):
+        s = StreamId(stream_id=256)
+
+    with pytest.raises(ValueError):
+        s = StreamId()
+
+    s = StreamId(device_id=5, stream_id=12)
+    assert s.__hash__() == 1292
+    assert s.__hash__() == (5 << 8) | 12
+
+
+def test_stream_packet_type():
+    s = StreamId(device_id=4, stream_id=8)
+    d = b"\x01\x02\x03\x04"
+    p = StreamPacket(s, d)
+    assert p
+    assert p.stream.device_id == 4
+    assert p._raw_data == b"\x01\x02\x03\x04"
+    assert p._raw_data_length == 4
+    assert p._byte_array == b"\x04\x08\x04\x00\x01\x02\x03\x04"
+    assert p.byte_array == b"\x04\x08\x04\x00\x01\x02\x03\x04"
+
+    d = bytearray([i for i in range(0xFF)])
+    p = StreamPacket(s, d)
+    assert p
+    assert p._raw_data_length == 0xFF
+    assert int.from_bytes(p.byte_array[2:4], "little") == 0xFF
+    assert p.byte_array[4:] == bytearray([i for i in range(0xFF)])
+
+
+def test_stream_manager_constructor():
+    sm = StreamManager()
+    assert sm
+    assert sm._lock.locked() == False
+    assert not sm._semaphore.locked() == False
+    assert len(sm._pending_data) == 0
+    assert len(sm._streams) == 0
+
+
+def test_stream_manager_adds_streams():
+    s0 = StreamId(device_id=0, stream_id=0)
+    s1 = StreamId(device_id=1, stream_id=0)
+
+    sm = StreamManager()
+
+    f0 = sm.add_stream(s0)
+    f1 = sm.add_stream(s1)
+    assert s0 in sm._streams
+    assert s1 in sm._streams
+    assert f0 != f1
+
+
+def test_stream_manager_removes_streams():
+    s0 = StreamId(device_id=0, stream_id=0)
+    s1 = StreamId(device_id=1, stream_id=0)
+
+    sm = StreamManager()
+
+    f0 = sm.add_stream(s0)
+    f1 = sm.add_stream(s1)
+    assert s0 in sm._streams
+    assert s1 in sm._streams
+    assert f0 != f1
+
+    sm.remove_stream(s0)
+    assert s0 not in sm._streams
+    assert s1 in sm._streams
+
+    sm.remove_stream(s1)
+    assert s0 not in sm._streams
+    assert s1 not in sm._streams

```

Comment:
Handle the case where the callbacks f0 and f1 are called after the stream has been removed?

Reply:
Sorta means that every device that has one of those callback functions has to drop it.  That could be as simple as seeing an exception next time it's called.

Reply:
Nah - the device will own that callback and shouldn't call it if the associated stream has been removed.